### ISSUE TITLE
Add unlifted versions of other primitive things

### ISSUE THREAD

**Issue by treeowl** [haskell-primitive/primitive-unlifted#14]
The ones that come to mind off the top of my head: `MVar`, `MutVar`, `Weak`, `StablePtr`, `TVar`, and `Ptr`. As I recall, there's a classic `MVar`-based concurrent queue implementation  that involves `MVar (MVar ...)`; we could offer an improved version.

**Comment by andrewthad** [comment]
Not `Ptr` and `StablePtr`. Those are both backed by `Addr#`. But all those other ones, yes.

**Comment by treeowl** [comment]
You're probably right about `Ptr`, but `StablePtr` is another story. That's backed by `StablePtr#`, which is a pointer into (or index in? Don't remember) the stable pointer table. It makes perfect sense to store unlifted things in that table, I believe.

**Comment by treeowl** [comment]
Oh, and `StableName` too! The type story there is stupid, so we would probably have something like

```haskell
-- No type argument!
newtype StableName#
  = StableName (Exts.StableName# Any)
```

then just wrap the primops with unsafe coercions to make them work with  unlifted things. We can make the same type work with lifted things while we're at it.

I believe another thing that makes sense is a version of `StableName` for singletons (or better, indexed by value in a `DependentHaskell` future), but that doesn't seem like a good fit for this package.

**Comment by andrewthad** [comment]
`StablePtr` cannot be made to work. We had to [remove the PrimUnlifted instance for StablePtr two years ago](https://github.com/haskell/primitive/pull/202), and nothing about the representation of `StablePtr#` has changed since then. You wouldn't even be able to write a `PrimUnlifted` instance for it without `unsafeCoerce#`, and `unsafeCoerce#` between types of boxed and unboxed representation is never sound.

**Comment by treeowl** [comment]
I think you're mixing up concerns. `StablePtr` can't have a `PrimUnlifted` instance, but we *can* have an `UnliftedStablePtr` that points to unlifted objects.

**Comment by andrewthad** [comment]
You're right. I had misread. Yes, that should work, and I've actually done before in an application, so I'm pretty confident that it does work.
